File: src/components/article/metadata.ts
import type {
  Update,
  Article,
  ArticleInfo,
  Frontmatter,
  FrontmatterUpdate,
} from './articleInfo'

export function toArticleInfo(article: Article): ArticleInfo {
  const stubArticleInfo: ArticleInfo = {
    url: '',
    file: '',
    title: 'ERROR, NO TITLE',
    date: new Date('2000-01-01'),
  }

  const metadata = article.frontmatter

  if (!metadata.title || !metadata.date) return stubArticleInfo
  const date = parseDate(metadata.date)
  const minImgUrl = getMinImgUrlOrFallbackToImgUrl(metadata)
  const updates = metadata.updates?.map(toArticleInfoUpdate)

  return {
    ...metadata,
    minImgUrl,
    date,
    updates,
    url: article.url,
    file: article.file,
  } as ArticleInfo
}

function getMinImgUrlOrFallbackToImgUrl(metadata: Frontmatter) {
  return metadata.minImgUrl === undefined || metadata.minImgUrl === ''
    ? metadata.imgUrl
    : metadata.minImgUrl
}

function toArticleInfoUpdate(update: FrontmatterUpdate): Update {
  return { diffUrl: update.diffUrl, date: parseDate(update.date) }
}

function parseDate(dateString: string): Date {
  return new Date(dateString)
}

File: renderer/markdown/parse/metadata.go
package parse

import (
	"html/template"
	"path/filepath"
	"strings"

	"github.com/blazejsewera/blog/renderer/article"
	"github.com/blazejsewera/blog/renderer/constants"
)

func mapToArticleMetadata(f rawFrontmatter, markdownFilename string) article.Metadata {
	sourceFile := markdownFilename
	url := urlFromMdFilename(markdownFilename)
	targetFile := targetFileFromMdFilename(markdownFilename)

	return article.Metadata{
		Title:            f.Title,
		Subtitle:         f.Subtitle,
		Date:             f.Date,
		Draft:            f.Draft,
		DraftDescription: f.DraftDescription,
		Author:           f.Author,
		Abstract:         f.Abstract,
		Keywords:         f.Keywords,
		Language:         f.Language,
		License:          f.License,
		ImgURL:           template.URL(f.ImgURL),
		ImgDescription:   f.ImgDescription,
		URL:              template.URL(url),
		SourceFile:       sourceFile,
		TargetFile:       targetFile,
		Updates:          toArticleUpdates(f.Updates),
	}
}

func toArticleUpdates(uu []update) []article.Update {
	if uu == nil {
		return nil
	}
	result := make([]article.Update, len(uu))
	for i, u := range uu {
		result[i] = u.toArticleUpdate()
	}
	return result
}

func (u update) toArticleUpdate() article.Update {
	return article.Update{
		Date:    u.Date,
		DiffURL: u.DiffURL,
	}
}

func urlFromMdFilename(markdownFilename string) string {
	htmlFilename := bareTarget(markdownFilename)
	posix := filepath.ToSlash(htmlFilename)
	return strings.TrimSuffix(posix, "/index.html")
}

func targetFileFromMdFilename(markdownFilename string) string {
	return constants.DistDir + bareTarget(markdownFilename)
}

func bareTarget(markdownFilename string) string {
	trimmedOfSite := strings.TrimPrefix(markdownFilename, constants.SiteDir)
	trimmedOfExt := strings.TrimSuffix(trimmedOfSite, constants.MdExt)
	return trimmedOfExt + constants.HtmlExt
}
